pydatastructs/trees/binary_trees.py

```diff
@@ -1053,3 +1054,132 @@ def get_sum(self, left_index, right_index):
+        super(SplayTree, self)._left_right_rotate(self.tree[p].parent, p)
+
+    def _zag_zag(self, x, p):
+        super(SplayTree, self)._left_rotate(self.tree[p].parent, p)
+        super(SplayTree, self)._left_rotate(p, x)
+
+    def _zag_zig(self, x, p):
+        super(SplayTree, self)._right_left_rotate(self.tree[p].parent, p)
+
+    def splay(self, x, p):
+        while self.tree[x].parent is not None:
+            if self.tree[p].parent is None:
+                self._zig(x, p)
+            elif self.tree[p].left == x and self.tree[self.tree[p].parent].left == p:
+                self._zig_zig(x, p)
+            elif self.tree[p].right == x and self.tree[self.tree[p].parent].right == p:
+                self._zag_zag(x, p)
+            elif self.tree[p].left == x and self.tree[self.tree[p].parent].right == p:
+                self._zag_zig(x, p)
+            else:
+                self._zig_zag(x, p)
+            p = self.tree[x].parent
+
+    def insert(self, key, x):
+        super(SelfBalancingBinaryTree, self).insert(key, x)
+        e, p = super(SelfBalancingBinaryTree, self).search(x, parent=True)
+        self.tree[self.size-1].parent = p;
+        self.splay(e, p)
+
+    def delete(self, x):
+        e, p = super(SelfBalancingBinaryTree, self).search(x, parent=True)
+        if e is None:
+            return
+        self.splay(e, p)
+        b = super(SelfBalancingBinaryTree, self).delete(x, balancing_info=True)
+        return True
+
+    def join(self, other):
+        """
+        Joins two trees self and other such that all elements of
+        the self tree are smaller than the elements of the other tree.
+
+        Parameters
+        ==========
+
+        other: SplayTree
+            SplayTree which needs to be joined with the self tree.
+
+        """
+        maxm = self.root_idx
+        while self.tree[maxm].right is not None:
+            maxm = self.tree[maxm].right
+        self.splay(maxm, self.tree[maxm].parent)
+        other.split_array(self, other.tree[other.root_idx].data)
+
+    def split(self, x):
+        """
+        Splits self tree into two trees such that one tree contains nodes
+        with key less than or equal to x and the other tree containing
+        nodes with key greater than x.
+
+        Parameters
+        ==========
+
+        x: key
+            Key of the element on the basis of which split is performed.
+
+        Returns
+        =======
+
+        other: SplayTree
+            SplayTree containing elements with key greater than x.
+
+        """
+        e, p = super(SelfBalancingBinaryTree, self).search(x, parent=True)
+        if e is None:
+            return
+        self.splay(e, p)
+        other = SplayTree(None, None)
+        if self.tree[self.root_idx].right is not None:
+            self.split_array(other, self.tree[self.tree[self.root_idx].right].data)
+            self.tree[self.root_idx].right = None
+        return other
+
+    def split_array(self, sec, x):
```

Comment:
Should this be public?

Reply:
Removed split_array function, as it was no longer useful in iterative algorithm using stacks.